### ISSUE TITLE
Provider config should not fall back to ambient config when explicit provider context is not set

### ISSUE THREAD

**Issue by mnlumi** [pulumi/pulumi-kubernetes#2357]
### What happened?

Given a Pulumi program with:
- Default provider for kubernetes set in Pulumi.<stack>.yaml
- An explicit provider defined with a kubeconfig containing multiple clusters
- No context specified to the defined provider
- A deployment attached to the provider

Even though default provider has been disabled, when the context of the explicit provider is not specified, the deployment will fall back to an ambient kubeconfig and context if available.

### Expected Behavior

I expect the ambient kubeconfig to <never> be used when the default providers are disabled in a stack and for the program to fail because the explicitly defined provider does not specify one context. The context should not be assumed. 

An equivalent situation with an unset context using kubectl results in failures.

### Steps to reproduce

1. Use a kubeconfig that has multiple contexts.
2. Disable default providers for the stack.
3. Have kubectl set to use a specific context.
4. Use the following program.

```
import pulumi
from pulumi_kubernetes.apps.v1 import Deployment, DeploymentSpecArgs
from pulumi_kubernetes.meta.v1 import LabelSelectorArgs, ObjectMetaArgs
from pulumi_kubernetes.core.v1 import ContainerArgs, PodSpecArgs, PodTemplateSpecArgs
from pulumi_kubernetes import Provider
from pulumi_kubernetes import yaml

app_labels = { "app": "nginx" }
## Load the kubeconfig file
kubeconfig = pulumi.Config("kubernetes").get_secret("kubeconfig")
# Create a Kubernetes provider using the kubeconfig
provider = Provider("provider", kubeconfig=kubeconfig) # <--No context specified

cg = yaml.ConfigGroup

deployment = Deployment(
    "nginx",
    spec=DeploymentSpecArgs(
        selector=LabelSelectorArgs(match_labels=app_labels),
        replicas=1,
        template=PodTemplateSpecArgs(
            metadata=ObjectMetaArgs(labels=app_labels),
            spec=PodSpecArgs(containers=[ContainerArgs(name="nginx", image="nginx")])
        ),
    ),opts=pulumi.ResourceOptions(provider=provider))

pulumi.export("name", deployment.metadata["name"])

```

### Output of `pulumi about`

```
CLI
Version      3.61.1
Go Version   go1.20.2
Go Compiler  gc

Plugins
NAME        VERSION
kubernetes  3.24.2
python      unknown

Host
OS       darwin
Version  13.0.1
Arch     arm64

Dependencies:
NAME               VERSION
pip                23.0.1
pulumi-kubernetes  3.24.2
setuptools         67.6.1
wheel              0.40.0
```

### Additional context

I have a repro and screen recording available. Have also discussed this with @lblackstone 

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by kpitzen** [comment]
Thanks for opening this @mnlumi - I agree that this seems like a bug (and quite a dangerous one at that). We'll want to prioritize this in an upcoming iteration. Thanks again!

**Comment by lblackstone** [comment]
@mnlumi I'm confused about your expectations here. For an explicit provider, if the provided kubeconfig has the `current-context` set, then the provider will use that context from the kubeconfig. However, if the provider's `context` parameter is set, it will override whatever is set in the kubeconfig. This behaves like I'd expect, since the context is explicitly set in both cases (either via the kubeconfig, or via a provider parameter).

Here's a concrete example:

kubeconfig
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <redacted>
    server: https://kubernetes.docker.internal:6443
  name: docker-desktop
contexts:
- context:
    cluster: docker-desktop
    user: docker-desktop
  name: docker-desktop
- context:
    cluster: invalid
    user: invalid
  name: invalid
current-context: docker-desktop
kind: Config
preferences: {}
users:
- name: docker-desktop
  user:
    client-certificate-data: <redacted>
    client-key-data: <redacted>
```

```ts
import * as k8s from "@pulumi/kubernetes";

const provider = new k8s.Provider("k8s", {
    kubeconfig: "/Users/levi/.kube/config",
    // context: "docker-desktop", <-- context not provided, so the `current-context` value is used from the specified kubeconfig
});
```

```ts
import * as k8s from "@pulumi/kubernetes";

const provider = new k8s.Provider("k8s", {
    kubeconfig: "/Users/levi/.kube/config",
    context: "invalid", // <-- value overrides the `current-context` value from the specified kubeconfig
});
```

A separate, but related concern is that the provider will load ambient kubeconfig settings if the `kubeconfig` parameter is not set on the explicit provider. I can see a case for changing that behavior for explicit providers.

```ts
import * as k8s from "@pulumi/kubernetes";

const provider = new k8s.Provider("k8s", {
    // kubeconfig: "/Users/levi/.kube/config", <-- not set, so the provider loads the default kubeconfig path
    // context: "docker-desktop", <-- context not provided, so the `current-context` value is used from the default kubeconfig
});
```

**Comment by rquitales** [comment]
@mnlumi Following up on this, I'm not sure if this is a behaviour change we'd want to commit as Levi explained. It makes sense to me as well that Pulumi should use the context set in the defined kubeconfig file if context is not explicitly provided to the provider as well. WDYT?

**Comment by lblackstone** [comment]
After further investigation, we've concluded that the current implementation is working as expected. Once a `kubeconfig` is selected, the provider will check the `current-context` setting in that file if `context` is not explicitly specified as an argument. That said, there are a couple of caveats that can lead to surprising results for users today:

1. If a `kubeconfig` is not specified, the client-go library in the provider uses the default lookup path to find one in a similar fashion to most Kubernetes CLI tooling. This can lead to a different `kubeconfig` being loaded depending on the state of the system running the Pulumi operation.
2. The `current-context` setting in the `kubeconfig` can change, which affects the context used by the provider if a `context` is not specified. If this setting is changed, then subsequent Pulumi operations will use the new context. 

Both of these settings can be explicitly set with Provider configuration, as well as disabling default providers, which avoids any ambiguity about the intended cluster and context during an operation.

I have opened https://github.com/pulumi/pulumi-kubernetes/issues/2423 to track a related enhancement, and am closing this issue since it is working as intended.
